File: curriculum/units/M11/code/typescript/harness.ts
export type InputSource = "repl" | "headless" | "tool";

export type TextBlock = {
  type: "text";
  text: string;
};

export type ToolUseBlock = {
  type: "tool_use";
  id: string;
  name: string;
  input: Record<string, unknown>;
};

export type ToolResultBlock = {
  type: "tool_result";
  toolUseId: string;
  content: unknown;
  isError: boolean;
};

export type UserMessage = {
  kind: "user";
  id: string;
  source: InputSource;
  blocks: Array<TextBlock | ToolResultBlock>;
};

export type AssistantMessage = {
  kind: "assistant";
  id: string;
  blocks: Array<TextBlock | ToolUseBlock>;
};

export type ConversationMessage = UserMessage | AssistantMessage;

export type ModelRequest = {
  id: string;
  messages: readonly ConversationMessage[];
};

export type TraceEvent = {
  type: string;
  detail?: Record<string, unknown>;
};

export class TraceSink {
  readonly events: TraceEvent[] = [];

  record(type: string, detail?: Record<string, unknown>): void {
    this.events.push(detail ? { type, detail } : { type });
  }
}

export class IdSource {
  private nextValue = 1;

  next(prefix: string): string {
    return `${prefix}-${this.nextValue++}`;
  }
}

export class SessionStore {
  private readonly messages: ConversationMessage[] = [];

  append(message: ConversationMessage): void {
    this.messages.push(structuredClone(message));
  }

  snapshot(): readonly ConversationMessage[] {
    return structuredClone(this.messages);
  }
}

export interface InputAdapter {
  readonly source: "repl" | "headless";
  accept(raw: string, ids: IdSource): UserMessage;
}

export class ReplInputAdapter implements InputAdapter {
  readonly source = "repl" as const;

  accept(raw: string, ids: IdSource): UserMessage {
    return {
      kind: "user",
      id: ids.next("message"),
      source: this.source,
      blocks: [{ type: "text", text: raw }],
    };
  }
}

export class HeadlessInputAdapter implements InputAdapter {
  readonly source = "headless" as const;

  accept(raw: string, ids: IdSource): UserMessage {
    return {
      kind: "user",
      id: ids.next("message"),
      source: this.source,
      blocks: [{ type: "text", text: raw }],
    };
  }
}

export class RequestProjector {
  private readonly ids: IdSource;

  constructor(ids: IdSource) {
    this.ids = ids;
  }

  project(messages: readonly ConversationMessage[]): ModelRequest {
    return {
      id: this.ids.next("request"),
      messages: structuredClone(messages),
    };
  }
}

export interface ModelAdapter {
  complete(request: ModelRequest, signal: AbortSignal): Promise<AssistantMessage>;
}

export interface Tool {
  readonly name: string;
  execute(input: Record<string, unknown>, signal: AbortSignal): Promise<unknown>;
}

export type RunResult = {
  status: "completed" | "cancelled" | "max_turns";
  finalMessage?: AssistantMessage;
};

export class AgentLoop {
  private readonly session: SessionStore;
  private readonly projector: RequestProjector;
  private readonly model: ModelAdapter;
  private readonly ids: IdSource;
  private readonly trace: TraceSink;
  private readonly maxTurns: number;

  constructor(
    session: SessionStore,
    projector: RequestProjector,
    model: ModelAdapter,
    tools: readonly Tool[],
    ids: IdSource,
    trace: TraceSink,
    maxTurns = 8,
  ) {
    this.session = session;
    this.projector = projector;
    this.model = model;
    this.tools = new Map(tools.map((tool) => [tool.name, tool]));
    this.ids = ids;
    this.trace = trace;
    this.maxTurns = maxTurns;
  }

  private readonly tools: Map<string, Tool>;

  async submit(
    rawInput: string,
    adapter: InputAdapter,
    signal: AbortSignal,
  ): Promise<RunResult> {
    const inputMessage = adapter.accept(rawInput, this.ids);
    this.session.append(inputMessage);
    this.trace.record("input.accepted", {
      source: adapter.source,
      messageId: inputMessage.id,
    });

    return this.run(signal);
  }

  private async run(signal: AbortSignal): Promise<RunResult> {
    for (let turn = 1; turn <= this.maxTurns; turn++) {
      if (signal.aborted) {
        this.trace.record("loop.cancelled", { phase: "before_request" });
        return { status: "cancelled" };
      }

      const request = this.projector.project(this.session.snapshot());
      this.trace.record("request.projected", {
        requestId: request.id,
        messageCount: request.messages.length,
      });
      this.trace.record("model.request", { requestId: request.id, turn });

      let assistant: AssistantMessage;
      try {
        assistant = await this.model.complete(request, signal);
      } catch (error) {
        if (signal.aborted || error instanceof CancelledError) {
          this.trace.record("loop.cancelled", { phase: "model" });
          return { status: "cancelled" };
        }
        throw error;
      }

      this.session.append(assistant);
      this.trace.record("assistant.received", { messageId: assistant.id });

      const toolUses = assistant.blocks.filter(
        (block): block is ToolUseBlock => block.type === "tool_use",
      );
      if (toolUses.length === 0) {
        this.trace.record("loop.completed", { turn });
        return { status: "completed", finalMessage: assistant };
      }

      for (let toolIndex = 0; toolIndex < toolUses.length; toolIndex++) {
        const call = toolUses[toolIndex]!;
        if (signal.aborted) {
          this.appendCancelledToolResults(
            toolUses.slice(toolIndex),
            "cancelled before execution",
          );
          this.trace.record("loop.cancelled", { phase: "before_tool" });
          return { status: "cancelled" };
        }

        this.trace.record("tool.started", {
          toolUseId: call.id,
          toolName: call.name,
        });
        const tool = this.tools.get(call.name);

        try {
          if (!tool) {
            throw new Error(`unknown tool: ${call.name}`);
          }
          const output = await tool.execute(call.input, signal);
          if (signal.aborted) {
            throw new CancelledError("cancelled during tool execution");
          }
          this.trace.record("tool.finished", {
            toolUseId: call.id,
            status: "success",
          });
          this.appendToolResult(call, output, false);
        } catch (error) {
          const cancelled = signal.aborted || error instanceof CancelledError;
          const content = cancelled
            ? "cancelled during tool execution"
            : error instanceof Error
              ? error.message
              : String(error);
          this.trace.record("tool.finished", {
            toolUseId: call.id,
            status: cancelled ? "cancelled" : "error",
          });
          this.appendToolResult(call, content, true);
          if (cancelled) {
            this.appendCancelledToolResults(
              toolUses.slice(toolIndex + 1),
              "cancelled before execution",
            );
            this.trace.record("loop.cancelled", { phase: "tool" });
            return { status: "cancelled" };
          }
        }
      }
    }

    this.trace.record("loop.max_turns", { maxTurns: this.maxTurns });
    return { status: "max_turns" };
  }

  private appendCancelledToolResults(
    calls: readonly ToolUseBlock[],
    content: string,
  ): void {
    for (const call of calls) {
      this.appendToolResult(call, content, true);
    }
  }

  private appendToolResult(
    call: ToolUseBlock,
    content: unknown,
    isError: boolean,
  ): void {
    const result: UserMessage = {
      kind: "user",
      id: this.ids.next("message"),
      source: "tool",
      blocks: [
        {
          type: "tool_result",
          toolUseId: call.id,
          content,
          isError,
        },
      ],
    };
    this.session.append(result);
    this.trace.record("tool_result.appended", {
      toolUseId: call.id,
      isError,
    });
  }
}

export class CancelledError extends Error {}

export class ScriptedModel implements ModelAdapter {
  readonly requests: ModelRequest[] = [];
  private readonly scripts: Array<
    (request: ModelRequest, callIndex: number) => AssistantMessage
  >;

  constructor(
    scripts: Array<
      (request: ModelRequest, callIndex: number) => AssistantMessage
    >,
  ) {
    this.scripts = scripts;
  }

  async complete(
    request: ModelRequest,
    signal: AbortSignal,
  ): Promise<AssistantMessage> {
    if (signal.aborted) throw new CancelledError("model cancelled");
    const copy = structuredClone(request);
    this.requests.push(copy);
    const script = this.scripts[this.requests.length - 1];
    if (!script) throw new Error("unexpected model request");
    return structuredClone(script(copy, this.requests.length - 1));
  }
}

export function findToolResult(
  messages: readonly ConversationMessage[],
  toolUseId: string,
): ToolResultBlock | undefined {
  for (const message of messages) {
    if (message.kind !== "user") continue;
    for (const block of message.blocks) {
      if (block.type === "tool_result" && block.toolUseId === toolUseId) {
        return block;
      }
    }
  }
  return undefined;
}

export function assistantWithText(id: string, text: string): AssistantMessage {
  return { kind: "assistant", id, blocks: [{ type: "text", text }] };
}

export function assistantWithTool(
  id: string,
  call: ToolUseBlock,
): AssistantMessage {
  return { kind: "assistant", id, blocks: [call] };
}

File: curriculum/units/M11/code/python/harness.py
from __future__ import annotations

import copy
from dataclasses import dataclass
from typing import Any, Awaitable, Callable, Protocol


Message = dict[str, Any]
ModelRequest = dict[str, Any]


class CancelledError(Exception):
    pass


class CancellationToken:
    def __init__(self) -> None:
        self.cancelled = False
        self.reason: str | None = None

    def cancel(self, reason: str) -> None:
        self.cancelled = True
        self.reason = reason

    def raise_if_cancelled(self) -> None:
        if self.cancelled:
            raise CancelledError(self.reason or "cancelled")


class IdSource:
    def __init__(self) -> None:
        self._next = 1

    def next(self, prefix: str) -> str:
        value = f"{prefix}-{self._next}"
        self._next += 1
        return value


class TraceSink:
    def __init__(self) -> None:
        self.events: list[dict[str, Any]] = []

    def record(self, event_type: str, **detail: Any) -> None:
        event: dict[str, Any] = {"type": event_type}
        if detail:
            event["detail"] = detail
        self.events.append(event)


class SessionStore:
    def __init__(self) -> None:
        self._messages: list[Message] = []

    def append(self, message: Message) -> None:
        self._messages.append(copy.deepcopy(message))

    def snapshot(self) -> list[Message]:
        return copy.deepcopy(self._messages)


class InputAdapter(Protocol):
    source: str

    def accept(self, raw: str, ids: IdSource) -> Message:
        ...


class ReplInputAdapter:
    source = "repl"

    def accept(self, raw: str, ids: IdSource) -> Message:
        return {
            "kind": "user",
            "id": ids.next("message"),
            "source": self.source,
            "blocks": [{"type": "text", "text": raw}],
        }


class HeadlessInputAdapter:
    source = "headless"

    def accept(self, raw: str, ids: IdSource) -> Message:
        return {
            "kind": "user",
            "id": ids.next("message"),
            "source": self.source,
            "blocks": [{"type": "text", "text": raw}],
        }


class RequestProjector:
    def __init__(self, ids: IdSource) -> None:
        self._ids = ids

    def project(self, messages: list[Message]) -> ModelRequest:
        return {
            "id": self._ids.next("request"),
            "messages": copy.deepcopy(messages),
        }


class ModelAdapter(Protocol):
    async def complete(
        self, request: ModelRequest, token: CancellationToken
    ) -> Message:
        ...


class Tool(Protocol):
    name: str

    async def execute(
        self, input_data: dict[str, Any], token: CancellationToken
    ) -> Any:
        ...


@dataclass(frozen=True)
class RunResult:
    status: str
    final_message: Message | None = None


class AgentLoop:
    def __init__(
        self,
        session: SessionStore,
        projector: RequestProjector,
        model: ModelAdapter,
        tools: list[Tool],
        ids: IdSource,
        trace: TraceSink,
        max_turns: int = 8,
    ) -> None:
        self._session = session
        self._projector = projector
        self._model = model
        self._tools = {tool.name: tool for tool in tools}
        self._ids = ids
        self._trace = trace
        self._max_turns = max_turns

    async def submit(
        self,
        raw_input: str,
        adapter: InputAdapter,
        token: CancellationToken,
    ) -> RunResult:
        input_message = adapter.accept(raw_input, self._ids)
        self._session.append(input_message)
        self._trace.record(
            "input.accepted",
            source=adapter.source,
            message_id=input_message["id"],
        )
        return await self._run(token)

    async def _run(self, token: CancellationToken) -> RunResult:
        for turn in range(1, self._max_turns + 1):
            if token.cancelled:
                self._trace.record("loop.cancelled", phase="before_request")
                return RunResult("cancelled")

            request = self._projector.project(self._session.snapshot())
            self._trace.record(
                "request.projected",
                request_id=request["id"],
                message_count=len(request["messages"]),
            )
            self._trace.record("model.request", request_id=request["id"], turn=turn)

            try:
                assistant = await self._model.complete(request, token)
            except CancelledError:
                self._trace.record("loop.cancelled", phase="model")
                return RunResult("cancelled")

            self._session.append(assistant)
            self._trace.record("assistant.received", message_id=assistant["id"])
            tool_uses = [
                block
                for block in assistant["blocks"]
                if block["type"] == "tool_use"
            ]

            if not tool_uses:
                self._trace.record("loop.completed", turn=turn)
                return RunResult("completed", assistant)

            for tool_index, call in enumerate(tool_uses):
                if token.cancelled:
                    self._append_cancelled_tool_results(
                        tool_uses[tool_index:], "cancelled before execution"
                    )
                    self._trace.record("loop.cancelled", phase="before_tool")
                    return RunResult("cancelled")

                self._trace.record(
                    "tool.started",
                    tool_use_id=call["id"],
                    tool_name=call["name"],
                )
                tool = self._tools.get(call["name"])

                try:
                    if tool is None:
                        raise RuntimeError(f"unknown tool: {call['name']}")
                    output = await tool.execute(call["input"], token)
                    token.raise_if_cancelled()
                    self._trace.record(
                        "tool.finished", tool_use_id=call["id"], status="success"
                    )
                    self._append_tool_result(call, output, is_error=False)
                except CancelledError:
                    self._trace.record(
                        "tool.finished", tool_use_id=call["id"], status="cancelled"
                    )
                    self._append_tool_result(
                        call, "cancelled during tool execution", is_error=True
                    )
                    self._append_cancelled_tool_results(
                        tool_uses[tool_index + 1 :], "cancelled before execution"
                    )
                    self._trace.record("loop.cancelled", phase="tool")
                    return RunResult("cancelled")
                except Exception as error:
                    self._trace.record(
                        "tool.finished", tool_use_id=call["id"], status="error"
                    )
                    self._append_tool_result(call, str(error), is_error=True)

        self._trace.record("loop.max_turns", max_turns=self._max_turns)
        return RunResult("max_turns")

    def _append_cancelled_tool_results(
        self, calls: list[dict[str, Any]], content: str
    ) -> None:
        for call in calls:
            self._append_tool_result(call, content, is_error=True)

    def _append_tool_result(
        self, call: dict[str, Any], content: Any, is_error: bool
    ) -> None:
        result = {
            "kind": "user",
            "id": self._ids.next("message"),
            "source": "tool",
            "blocks": [
                {
                    "type": "tool_result",
                    "toolUseId": call["id"],
                    "content": content,
                    "isError": is_error,
                }
            ],
        }
        self._session.append(result)
        self._trace.record(
            "tool_result.appended", tool_use_id=call["id"], is_error=is_error
        )


Script = Callable[[ModelRequest, int], Message]


class ScriptedModel:
    def __init__(self, scripts: list[Script]) -> None:
        self._scripts = scripts
        self.requests: list[ModelRequest] = []

    async def complete(
        self, request: ModelRequest, token: CancellationToken
    ) -> Message:
        token.raise_if_cancelled()
        request_copy = copy.deepcopy(request)
        self.requests.append(request_copy)
        index = len(self.requests) - 1
        if index >= len(self._scripts):
            raise RuntimeError("unexpected model request")
        return copy.deepcopy(self._scripts[index](request_copy, index))


def find_tool_result(messages: list[Message], tool_use_id: str) -> dict[str, Any] | None:
    for message in messages:
        if message["kind"] != "user":
            continue
        for block in message["blocks"]:
            if block["type"] == "tool_result" and block["toolUseId"] == tool_use_id:
                return block
    return None


def assistant_with_text(message_id: str, text: str) -> Message:
    return {
        "kind": "assistant",
        "id": message_id,
        "blocks": [{"type": "text", "text": text}],
    }


def assistant_with_tool(
    message_id: str, tool_use_id: str, name: str, input_data: dict[str, Any]
) -> Message:
    return {
        "kind": "assistant",
        "id": message_id,
        "blocks": [
            {
                "type": "tool_use",
                "id": tool_use_id,
                "name": name,
                "input": input_data,
            }
        ],
    }
